include/boost/beast/http/impl/basic_parser.hpp

```diff
@@ -56,6 +56,16 @@ put(ConstBufferSequence const& buffers,
         buf_.get(), size}, ec);
 }
 
+template<bool isRequest>
+boost::optional<std::uint64_t>
+basic_parser<isRequest>::
+impl_content_length() const
```

Comment:
How is this different from the already existing function `basic_parser::content_length`?

Reply:
The public function cannot be called during parsing because the assert will fail. So I have split it up into a public function which asserts the precondition and a private one that does not.

Reply:
The problem is that users can subclass `basic_parser`, so you have given `parser` a capability that ordinary users do not have access to (since `impl_content_length` is private).

Reply:
The capabilities and behaviour of the public API have not changed at all.
I have merely refactored the public function’s implementation to be expressed in terms of a private function which is now used in the parsing of headers (where calling the public API would assert)